codegen/facelift/templates/IPCServiceAdapter.template.cpp

```diff
@@ -109,42 +100,69 @@ void {{className}}::appendDBUSIntrospectionData(QTextStream &s) const
 {
     Q_UNUSED(s);   // For empty interfaces
     {% for property in interface.properties %}
-    ::facelift::DBusSignatureHelper::addPropertySignature<ServiceType::PropertyType_{{property.name}}>(s, "{{property.name}}", {{ property.readonly | cppBool }});
+    {% if not property.type.is_model %}
+    s << QStringLiteral("<property name=\"{{property.name}}\" type=\"%1\" access=\"%2\"/>").arg(typeToSignature<{{property.type.interfaceCppType}}>(),
+                               {{ property.readonly | cppBool }} ? QStringLiteral("read"): QStringLiteral("readwrite"));
+    {% endif %}
     {% endfor %}
+    s << QStringLiteral("<property name=\"ready\" type=\"b\" access=\"read\"/>");
+
     {% for operation in interface.operations %}
+    s << "<method name=\"{{operation.name}}\">";
+    {%- for parameter in operation.parameters -%}
+    s << "<arg name=\"{{parameter.name}}\" type=\"";
+    s << typeToSignature<{{parameter.type.interfaceCppType}}>();
+    s << "\" direction=\"in\"/>";
+    {%- endfor -%}
 
-    {
-        std::array<const char*, {{ operation.parameters.__len__() }}> argumentNames = { {
-            {%- for parameter in operation.parameters -%}
-            "{{parameter}}",
-            {%- endfor -%}
-        } };
-        ::facelift::DBusSignatureHelper::addMethodSignature<
-        {%- set comma = joiner(", ") -%}
-        {%- for parameter in operation.parameters -%}
-            {{ comma() }}{{parameter.cppType}}
-        {%- endfor -%}
-        >(s, "{{operation.name}}", argumentNames);
-    }
+    {% if operation.hasReturnValue %}
+    s << "<arg name=\"returnValue\" type=\"";
+    s << typeToSignature<{{operation.interfaceCppType}}>();
+    s << "\" direction=\"out\"/>";
+    {% endif %};
+    s << "</method>";
     {% endfor %}
 
     // signals
     {% for signal in interface.signals %}
     {
-        std::array<const char*, {{ signal.parameters.__len__() }}> argumentNames = { {
-            {%- for parameter in signal.parameters -%}
-                "{{parameter}}",
-            {%- endfor -%}
-        }};
-        ::facelift::DBusSignatureHelper::addSignalSignature<
-        {%- set comma = joiner(", ") -%}
+        s << "<signal name=\"{{signal.name}}\">";
         {%- for parameter in signal.parameters -%}
-        {{ comma() }}{{parameter.interfaceCppType}}
+        s << "<arg name=\"{{parameter.name}}\" type=\"";
+        s << typeToSignature<{{parameter.type.interfaceCppType}}>();
+        s << "\" direction=\"out\"/>";
         {%- endfor -%}
-        >(s, "{{signal.name}}", argumentNames);
+        s << "</signal>";
     }
-
     {% endfor %}
+
+    {% if interface.hasModelProperty %}
+    s << "<signal name=\"" << facelift::IPCCommon::MODEL_DATA_CHANGED_MESSAGE_NAME << "\">";
```

Comment:
that code is generic so it should not be duplicated.

Reply:
The whole "model" should be treated as separate interface (similar to org.freedesktop.DBus.Properties with its own signals and methods) so these will be improved automatically.